File: Palindrome Partitioning.cpp
class Solution {
public:
    string hash(vector<string> v) {
        stringstream ss;
        for (string& s: v) {
            ss << s << "-";
        }
        return ss.str();
    }
    vector<vector<string>> partition(string s) {
        int j = 0;
        vector<string> substring;
        for (char& c: s) {
            string new_s(1, c);
            substring.push_back(new_s);
        }

        vector<vector<string>> arr{
            substring,
        };

        unordered_set<string> seen;
        function<void(vector<string>)> explore;
        explore = [&arr, &seen, this] (vector<string> arr2) -> void {
            int n = arr2.size();
            for (int i = 0; i < n-1; i++) {
                if (arr2[i] == arr2[i+1]) {
                    vector<string> arr3(arr2.begin(), arr2.begin()+i);
                    arr3.push_back(arr2[i] + arr2[i+1]);
                    arr3.insert(arr3.end(), arr2.begin()+i+2, arr2.end());
                    string key = hash(arr3);
                    if (!seen.contains(key)) {
                        seen.insert(key);
                        arr.push_back(move(arr3));
                    }
                }
            }

            for (int i = 0; i < n-2; i++) {
                if (arr2[i] == arr2[i+2]) {
                    vector<string> arr3(arr2.begin(), arr2.begin()+i);
                    arr3.push_back(arr2[i] + arr2[i+1] + arr2[i+2]);
                    arr3.insert(arr3.end(), arr2.begin()+i+3, arr2.end());
                    string key = hash(arr3);
                    if (!seen.contains(key)) {
                        seen.insert(key);
                        arr.push_back(move(arr3));
                    }
                }
            }
        };

        while (j < arr.size()) {
            explore(arr[j++]);
        }
        return arr;
    }
};

File: Palindrome Partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        i = 0
        arr = [list(s)]
        seen = set()

        def explore(arr2):
            n = len(arr2)
            for i in range(n - 1):
                if arr2[i] == arr2[i + 1]:
                    arr3 = arr2[:i] + [arr2[i] + arr2[i + 1]] + arr2[i + 2:]
                    key = str(arr3)
                    if key not in seen:
                        seen.add(key)
                        arr.append(arr3)

            for i in range(n - 2):
                if arr2[i] == arr2[i + 2]:
                    arr3 = arr2[:i] + [arr2[i] + arr2[i + 1] + arr2[i + 2]] + arr2[i + 3:]
                    key = str(arr3)
                    if key not in seen:
                        seen.add(key)
                        arr.append(arr3)

        while i < len(arr):
            explore(arr[i])
            i += 1
        return arr
